browser/de_amp/de_amp_browsertest.cc

```diff
@@ -0,0 +1,278 @@
+  }
+
+  void SetUpInProcessBrowserTestFixture() override {
+    InProcessBrowserTest::SetUpInProcessBrowserTestFixture();
+    mock_cert_verifier_.SetUpInProcessBrowserTestFixture();
+  }
+
+  void TearDownInProcessBrowserTestFixture() override {
+    InProcessBrowserTest::TearDownInProcessBrowserTestFixture();
+    mock_cert_verifier_.TearDownInProcessBrowserTestFixture();
+  }
+
+  de_amp::DeAmpService* service() {
+    return de_amp::DeAmpServiceFactory::GetForBrowserContext(
+        web_contents()->GetBrowserContext());
+  }
+
+  content::WebContents* web_contents() {
+    return browser()->tab_strip_model()->GetActiveWebContents();
+  }
+
+  void TogglePref(const bool on) {
+    service()->ToggleDeAmp(on);
+    web_contents()->GetController().Reload(content::ReloadType::NORMAL, false);
+  }
+
+  void NavigateToURLAndWaitForRedirects(const GURL& original_url,
+                                        const GURL& landing_url) {
+    ui_test_utils::UrlLoadObserver load_complete(
+        landing_url, content::NotificationService::AllSources());
+    ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), original_url));
+    load_complete.Wait();
+    EXPECT_EQ(web_contents()->GetLastCommittedURL(), landing_url);
+  }
+
+  void GoBack(Browser* browser) {
+    content::TestNavigationObserver observer(web_contents());
+    chrome::GoBack(browser, WindowOpenDisposition::CURRENT_TAB);
+    observer.Wait();
+  }
+
+  void GoForward(Browser* browser) {
+    content::TestNavigationObserver observer(web_contents());
+    chrome::GoForward(browser, WindowOpenDisposition::CURRENT_TAB);
+    observer.Wait();
+  }
+
+ protected:
+  base::test::ScopedFeatureList feature_list_;
+  std::unique_ptr<net::EmbeddedTestServer> https_server_;
+  content::ContentMockCertVerifier mock_cert_verifier_;
+};
+
+std::unique_ptr<net::test_server::HttpResponse> HandleRequest(
+    const std::string& base_url,
+    const std::string& canonical_page,
+    const std::string& body,
+    const net::test_server::HttpRequest& request) {
+  auto port = request.GetURL().port();
+
+  auto http_response = std::make_unique<net::test_server::BasicHttpResponse>();
+  http_response->set_code(net::HTTP_OK);
+  http_response->set_content_type("text/html");
+  // ... rel="canonical" href="https://%s:%s%s" becomes
+  // href="https://<test server base url>:<test port>/<canonical html page>"
+  http_response->set_content(base::StringPrintf(
+      body.c_str(), base_url.c_str(), port.c_str(), canonical_page.c_str()));
+  return http_response;
+}
+
+IN_PROC_BROWSER_TEST_F(DeAmpBrowserTest, SimpleDeAmp) {
+  TogglePref(true);
+  ASSERT_TRUE(https_server_->Start());
+
+  // Go to any page
+  GURL simple = https_server_->GetURL(kTestHost, kTestSimpleNonAmpPage);
+  ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), simple));
+  EXPECT_EQ(web_contents()->GetLastCommittedURL(), simple);
+
+  // Now go to an AMP page
+  https_server_.reset(new net::EmbeddedTestServer(
```

Comment:
Is it necessary to recreate the server here?

Reply:
Yes, in order to register the request handler